src/main/kotlin/com/openlattice/indexing/BackgroundLinkingIndexingService.kt

```diff
@@ -289,48 +305,53 @@ class BackgroundLinkingIndexingService(
-     * Returns the linking ids, which are needing to be indexed along with their last_write and entity sets.
+     * Returns the linking ids, which are needing to be indexed along with last_write and their entity set ids and
+     * origin ids as a string in the format: "entity_set_id|origin_id".
      * Either because of property change or because of partial entity deletion(soft or hard) from the cluster.
      */
-    private fun getDirtyLinkingIds(): BasePostgresIterable<Triple<UUID, OffsetDateTime, Set<UUID>>> {
+    private fun getDirtyLinkingIds(): BasePostgresIterable<Triple<List<String>, UUID, OffsetDateTime>> {
         return BasePostgresIterable(
                 StatementHolderSupplier(hds, selectDirtyLinkingIds, FETCH_SIZE)
-        ) {
+        ) { rs ->
+            val entityDataKeysRaw = PostgresArrays.getTextArray(rs, ENTITY_DATA_KEY).toList()
             Triple(
-                    ResultSetAdapters.linkingId(it),
-                    ResultSetAdapters.lastWriteTyped(it),
-                    ResultSetAdapters.entitySetIds(it)
+                    entityDataKeysRaw,
+                    ResultSetAdapters.linkingId(rs),
+                    ResultSetAdapters.lastWriteTyped(rs)
             )
         }
     }
 
 
     /**
-     * Returns the linking ids along with last_write and its entity set ids, which are needing to be un-indexed
-     * (to delete those documents).
+     * Returns the linking ids which are needing to be un-indexed (to delete those documents) along with last_write and
+     * their entity set ids and origin ids as a string in the format: "entity_set_id|origin_id".
      */
-    private fun getDeletedLinkingIds(): BasePostgresIterable<Triple<UUID, OffsetDateTime, Set<UUID>>> {
+    private fun getDeletedLinkingIds(): BasePostgresIterable<Triple<List<String>, UUID, OffsetDateTime>> {
         return BasePostgresIterable(
                 StatementHolderSupplier(hds, selectDeletedLinkingIds, FETCH_SIZE)
-        ) {
+        ) { rs ->
+            val entityDataKeysRaw = PostgresArrays.getTextArray(rs, ENTITY_DATA_KEY).toList()
             Triple(
-                    ResultSetAdapters.linkingId(it),
-                    ResultSetAdapters.lastWriteTyped(it),
-                    ResultSetAdapters.entitySetIds(it)
+                    entityDataKeysRaw,
+                    ResultSetAdapters.linkingId(rs),
+                    ResultSetAdapters.lastWriteTyped(rs)
             )
         }
     }
 }
 
+internal const val ENTITY_DATA_KEY = "entity_data_key"
+
 /**
  * Select linking ids, where ALL normal entities are cleared or deleted.
  */
 internal val selectDeletedLinkingIds =
         // @formatter:off
         "SELECT " +
                 "${LINKING_ID.name}, " +
-                "max(${LAST_WRITE.name}) AS ${LAST_WRITE.name}, " +
-                "array_agg(${ENTITY_SET_ID.name}) AS ${ENTITY_SET_IDS.name} " +
+                "array_agg(${ENTITY_SET_ID.name} || '|' || ${ID.name}) AS $ENTITY_DATA_KEY, " +
```

Comment:
why are we selecting the values like this (joined with "|") instead of just selecting ..., entity_set_id, id, ... (since we just end up having to split it later anyway)

Reply:
Because we need all the entity data keys/linking_id, otherwise, when indexing the linking entity, not all data will be included. There is a fetch_size on the query, so even if we group by linking id/entity set id/id and then map it to a map, the last entries probably won't be complete. 
We could do 1 array creation (array[entity_set_id, id]) and in an outer query an array aggregations instead of this, if that is the preferred solution, where the result would be an array of entity set id-id pairs in an array. 
I choose to select also ids in these queries, so that we don't need to make an extra query to get all (normal)ids/entity set ids for each linking entity.

Reply:
The above doesn't quite compute for me. Are you saying that it is not possible to perform a partial update of a linked entity and you have to index the entire linked entity?

Reply:
Partial update probably would be possible, although I think it would be much more complicated then just querying the whole linked entity, but there is the additional problem of not knowing when to mark it as indexed, so when did we include all values?